### ISSUE TITLE
Plugins should not blindly overwrite schemas generated by other plugins

### ISSUE THREAD

**Issue by ilx** [marshmallow-code/apispec#592]
Discovered when we tried to combine marshmallow plugin with one-of marshmallow plugin in scope of flask-smorest lib:

```
pip install flask marshmallow marshmallow-dataclass marshmallow-enum marshmallow-oneofschema apispec apispec-oneofschema apispec-webframeworks
```
then the following script creates different output for field `type` based on the order of plugins:

```python
from dataclasses import dataclass, field
from enum import unique, Enum
from typing import ClassVar, Type

from apispec import APISpec
from apispec.ext.marshmallow import MarshmallowPlugin
from apispec_webframeworks.flask import FlaskPlugin
from apispec_oneofschema import MarshmallowPlugin as OneOfMarshmallowPlugin
import marshmallow
from marshmallow_dataclass import add_schema


@unique
class RepoType(Enum):
    GIT = 1
    GITHUB = 2
    GITLAB = 3


@add_schema
@dataclass
class Repository:
    Schema: ClassVar[Type[marshmallow.Schema]] = marshmallow.Schema

    class Meta:
        ordered = True

    name: str = field()
    type: RepoType = field(metadata={"by_value": False})


repo_schema = Repository.Schema()

flask_plugin = FlaskPlugin()
ma_plugin = MarshmallowPlugin()
one_of_plugin = OneOfMarshmallowPlugin()

spec = APISpec(
    title="Gisty",
    version="1.0.0",
    openapi_version="3.0.2",
    info=dict(description="A minimal gist API"),
    plugins=[
        flask_plugin,
        ma_plugin,
        one_of_plugin,
    ],
)


def enum2properties(self, field, **kwargs):
    import marshmallow_enum

    """Add an OpenAPI extension for marshmallow_enum.EnumField instances
    """
    ret = {}
    if isinstance(field, marshmallow_enum.EnumField):
        values = []
        for member in field.enum:
            values.append(member.name)

        ret["type"] = "string"
        ret["enum"] = values  # type: ignore
    return ret


ma_plugin.converter.add_attribute_function(enum2properties)
# one_of_plugin.converter.add_attribute_function(enum2properties)

spec.components.schema("Repository", schema=repo_schema)

print(spec.to_yaml())
```

so one output would be:
```yaml
components:
  schemas:
    Repository:
      properties:
        name:
          type: string
        type: {}
      required:
      - name
      - type
      type: object
info:
  description: A minimal gist API
  title: Gisty
  version: 1.0.0
openapi: 3.0.2
paths: {}
```

while if we re-order plugins we would get:

```yaml
components:
  schemas:
    Repository:
      properties:
        name:
          type: string
        type:
          enum:
          - GIT
          - GITHUB
          - GITLAB
          type: string
      required:
      - name
      - type
      type: object
info:
  description: A minimal gist API
  title: Gisty
  version: 1.0.0
openapi: 3.0.2
paths: {}
```

In the example above we added enum2properties attribute function only to the marshmallow plugin - as we expected one_of_plugin would not have impact on it, BUT inside apispec.core.Components.schema:

```python
        for plugin in self._plugins:
            try:
                ret.update(plugin.schema_helper(name, component, **kwargs) or {})
            except PluginMethodNotImplementedError:
                continue
        self._schemas[name] = ret
```

basically each plugin could blindly overwrite ret dict if it returns partial data (ie even if it does not know how to handle some field).

Now, we figured out that we could simply add attribute function to the one_of plugin (as it extends marshmallow plugin), but still... this could be handled in a bit more graceful way (maybe even issue a warning?).

Since both plugins are actually marshmallow plugins we thought we could simply use just one_of plugin, but it turns out that it produces yet another error when we try that (inside apispec.ext.marshmallow.openapi.OpenAPIConverter.get_ref_dict) - it's this one: https://github.com/timakro/apispec-oneofschema/issues/8

**Comment by Bangertm** [comment]
Just submitted a [pull request](https://github.com/timakro/apispec-oneofschema/pull/9) to apispec_oneofschema to resolve the issue with that library. So you should be able to fix the issue by only using a single marshmallow plugin. In general there should not be two plugins registered to do exactly the same thing so I'm not sure that apispec should be worried about the overwrite case.

**Comment by lafrech** [comment]
Let's keep things simple for now. We'll reopen if another clash case comes out.
